### PR TITLE
Upgrade to electron 5.0.12

### LINKED ISSUE #20034: Crashing on macOS 10.15
### Prerequisites

* [ x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: <https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode>
    * Followed all applicable steps in the debugging guide: <https://flight-manual.atom.io/hacking-atom/sections/debugging/>
    * Checked the FAQs on the message board for common solutions: <https://discuss.atom.io/c/faq>
    * Checked that your issue isn't already filed: <https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom>
    * Checked that there is not already an Atom package that provides the described functionality: <https://atom.io/packages>

### Description

Atom has been crashing on me ever since upgrading to macOS 10.15 (Catalina). Has never really crashed on me (that I can remember) before Catalina. Now it does it a few times daily. Randomly. Sometimes when deleting a folder or file,  duplicating a file, renaming a file, moving a file, and toggling a directory in the file tree. I tried to look in the crash log, but there was none. This the only message I could find related to Atom:

MacBook-Air Atom[11027]: assertion failed: 19A602: libxpc.dylib + 86780 [2E9076CD-6C0E-38B6-8403-B2DDCE125FBF]: 0x89

### Steps to Reproduce

Interacting with the file tree (toggle folders, moving a file, renaming, deleting, duplicating, trying to search in a directory, saving, ect..)

**Expected behavior:**

to interact with file tree without crashing the whole application, generate crash log in console

**Actual behavior:**

Atom will close at random times during the day only when interacting with file tree. A couple times it had closed when saving a file.

**Reproduces how often:**

It happens randomly, daily, I guess the more you use file tree you can expect more times. 

### Versions

Atom    : 1.40.1
Electron: 3.1.10
Chrome  : 66.0.3359.181
Node    : 10.2.0

<hr>

**Update:**

Here's a temporary patch that resolves the crashing, replaces the Electron context-menu with a HTML version:
[Atom Menu Patch for Catalina](https://github.com/facebookexperimental/atom-menu-patch-for-catalina)
You can also search and install the package inside Atom.

### PR DESCRIPTION
**Pull request by darangi**
This upgrades atom from `4.2.7` to `5.0.12`, this helps prevent the crash happening on Mac OS 10.15 (Catalina)

# TODO

- [x] Update CI config to use `node v12.13.1` and `npm v6.12.1`
- [x] Get CI passing
- [x] Run through Atom's Electron upgrade test manifest (below) to identify any regressions

# ISSUES TO RESOLVE

- [x] Bump `leveldown@4.0.0` to `leveldown@5.4.1` in  `electron-link`  to fix build error 
- [x] publish a new version for `electron-link@0.4.1`  https://github.com/atom/electron-link
- [x] Bump `oniguruma@6.2.1` to `oniguruma@7.2.1` in `apm` to fix build error
- [x] publish a new  version for `apm@2.4.4`  version  https://github.com/atom/apm 
- [x] Fix git-utils build error https://github.com/atom/git-utils/pull/98/
  - [x] publish a new version for `git-utils@5.6.2` https://github.com/atom/git-utils
  - [x] Bump `git-utils@5.6.1` to  `git-utils@5.6.2` in `atom/atom` and `apm` to fix build error 
- [x] Fix fuzzy-native build error  https://github.com/atom/fuzzy-native/pull/8 thrown from `fuzzy-finder` https://github.com/atom/fuzzy-finder
   - [x] publish a new version for `fuzzy-native@1.1.2` https://github.com/atom/fuzzy-native/releases/tag/v1.1.2
  - [x] Bump `fuzzy-native@1.1.1` to  `fuzzy-native@1.1.2` in `fuzzy-finder` to fix build error https://github.com/atom/fuzzy-finder/pull/409
   - [x] publish a new version for `fuzzy-finder@1.14.2` https://github.com/atom/fuzzy-finder/releases/tag/v1.14.2
   - [x] Bump `fuzzy-finder@1.14.1` to  `fuzzy-finder@1.14.2` in `atom/atom` to fix build error 
- [x] Bump `fs-admin@0.5.0` to `fs-admin 0.12.0` to fix build error 
- [x] Fix compilation errors thrown from `text-buffer` https://github.com/atom/text-buffer/pull/317 
- [x] Resolve https://github.com/atom/node-ctags/issues/16, (https://github.com/atom/node-ctags/pull/17)
- [x] Resolve https://github.com/tree-sitter/tree-sitter-c/issues/35
- [x] Resolve https://github.com/tree-sitter/tree-sitter-css/issues/10
   - [x] Bump `tree-sitter-css@0.13.7` to `tree-sitter-css@0.13.8` in `language-css` to fix build error
   - [x] Publish a new version of `language-css` https://github.com/atom/language-css/releases/tag/v0.44.4
- [x] Resolve https://github.com/tree-sitter/tree-sitter-embedded-template/issues/3 (https://github.com/tree-sitter/tree-sitter-embedded-template/pull/4)
   - [x] Bump `tree-sitter-embedded-template@0.15.0` to `tree-sitter-embedded-template@0.15.2` in `language-html` to fix build error
  - [x] Publish a new version for `language-html@0.53.1` https://github.com/atom/language-html/releases/tag/v0.53.1
- [x] Bump `tree-sitter-go@0.15.0` to `tree-sitter-go@0.15.1` in `language-go` to fix build error
  - [x] Publish a new version for `language-go@0.47.1` https://github.com/atom/language-go/releases/tag/v0.47.2
- [x] Publish new patch for `tree-sitter-javascript@0.15.2` https://github.com/tree-sitter/tree-sitter-javascript/releases/tag/v0.15.2
  - [x] Bump `tree-sitter-javascript@0.15.0` to `tree-sitter-javascript@0.15.2` in `language-javascript` to fix build error.
  - [x] Publish a new version for `language-javascript@0.133.2` https://github.com/atom/language-javascript/releases/tag/v0.134.1
- [x] Bump `tree-sitter-json@0.14.0` to `tree-sitter-json@0.15.1` in `language-json` to fix build error
  - [x] Publish a new version for `language-json@1.0.5` https://github.com/atom/language-json/releases/tag/v1.0.5
- [x] Bump `tree-sitter-python@0.15.0` to `tree-sitter-json@0.15.1` in `language-python` to fix build error
  - [x] Publish a new version for `language-python@0.53.4` https://github.com/atom/language-python/releases/tag/v0.53.4
- [x] Publish new patch for `tree-sitter-typescript@0.15.2` https://github.com/tree-sitter/tree-sitter-typescript/releases/tag/v0.15.2
  - [x] Bump `tree-sitter-typescript@0.15.1` to `tree-sitter-typescript@0.15.2` in `language-typescript` to fix build error.
  - [x] Publish a new version for `language-typescript@0.6.1` https://github.com/atom/language-typescript/releases/tag/v0.6.1
- [x] Bump `language-java@0.31.3` to `language-java@0.13.4` in `atom/atom` 
- [x] Bump `tree-sitter-rust@0.15.1` to `tree-sitter-rust@0.15.1` in `language-rust-bundled` to fix build errors
- [x] Fix errors thrown from github package 
  - [x] Resolve https://github.com/kuychaco/whats-my-line/issues/6
  - [x] Bump `whats-my-line@0.1.1` to `whats-my-line@0.1.2` in `github`
- [x] Resolve https://github.com/atom/atom/issues/20236
- [x] Resolve `chrome-sandbox` bug on linux https://github.com/atom/atom/issues/20289 (https://github.com/atom/atom/pull/20172/commits/18da63097542d1904e34483b491099f42ce12575)
- [x] Resolve https://github.com/atom/atom/issues/20306 (https://github.com/atom/atom/pull/20172/commits/47154351f0de65e89aed6fb4a35a392d952832e7)
- [x] Resolve Tests timing out on CI https://github.com/atom/atom/issues/20290 (https://github.com/atom/atom/pull/20172/commits/1bc06505d5cccb452724c5628f6288787f16aa44)
- [x] Resolve Failed to load a language-rust-bundled package grammar  https://github.com/atom/atom/issues/20354 (https://github.com/atom/atom/pull/20172/commits/67d61c035134311ea12da5153bb2c556a6a3b911)
- [x] Resolve Failed to load a language-shellscript package grammar https://github.com/atom/atom/issues/20353 (https://github.com/atom/atom/pull/20172/commits/2a9c63d5b783ee90a6b271e4661b5ccec8d06708)
- [x] Resolve Failed to load a language-typescript package grammar https://github.com/atom/atom/issues/20355 (https://github.com/atom/atom/pull/20172/commits/081dd6019b30a67c6f044c4b113fa719aedf37eb)

# FAILING TESTS
- [x] TreeSitterLanguageMode https://github.com/atom/atom/issues/20371
- [x] Failing tests in Reporter package https://github.com/atom/atom/issues/20399
- [x] Failing tests in snippet package https://github.com/atom/atom/issues/20386
- [x] Failing tests in find and replace package https://github.com/atom/atom/issues/20376
- [x]  GIthub package tests exits on CI https://github.com/atom/atom/issues/20411
# Regressions to fix before merging

# Known bugs resolved by this upgrade

Fixes https://github.com/atom/atom/issues/20034

# Electron Upgrade Test Manifest

- [x] Review issues tagged with [`electron`](https://github.com/search?utf8=%E2%9C%93&q=org%3Aatom+label%3Aelectron+is%3Aopen+is%3Aissue) to see if any of them are fixed
- [x] Verify that you can collaborate with Teletype
    - [x] Host using Atom's current stable release can collaborate with guest using Atom with the new version of Electron
    - [x] Guest using Atom's current stable release can collaborate with host using Atom with the new version of Electron
    - [x] Host and guest can collaborate when both are using Atom with the new version of Electron
    <details>
    <summary>Steps</summary>

    1. Open a file
    1. Share a portal
    1. Have a guest join the portal
    1. When guest edits the text in the file, verify that the host sees those changes
    </details>
### Text Input/Keybindings

See how to [setup keyboard layouts](https://github.com/atom/maintainers-log/blob/master/how-tos/setup-keyboard-layouts.md).

- [x] IME not working (https://github.com/atom/atom/issues/14911)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Pinyin simplified  
  1. Open Atom and use the layout you installed as the keyboard layout
  1. Example, if you picked Pinyin simplified type in `zhongwenshuru`

  **Expected:** in Pinyin simplified `中文输入` is expected

  gif

  **Issues we've seen:** Only shows latin first character and not every character typed.

  gif
  </details>

- [x] macOS: Key binding with composition characters are ignored (https://github.com/atom/atom/issues/15189)
  <details>
  <summary>Repro steps</summary>

  1. Open your keymap.cson file and add the following:
      ```cson
      'atom-text-editor':
        # use hjkl; vim-like when pressing alt (word by word, paragraph by paragraph)
        # or alt (letter by letter)
        'alt-h': 'core:move-left'
        'alt-j': 'core:move-down'
        'alt-k': 'core:move-up'
        'alt-l': 'core:move-right'
      ```
  1. Save keymap.cson and reload atom.
  1. Try the new keymappings under `ABC - Extended` keyboard layout

  **Expected**: the keybindings work

  **Issues we've seen:** it types out the modified keys and ignores the key mapping
  </details>

- [x] macOS: Composition characters mess up insertion point (https://github.com/atom/atom/issues/15344)
  <details>
  <summary>Repro steps</summary>

  1. Open keymap.cson and add the following:
      ```cson
      'atom-text-editor':
        'alt-n': 'core:move-down'
        'alt-o': 'core:move-down'
      ```
  1. Have the following in your buffer:
      ```
      Foo
      Bar
      ```
  1. Have the cursor at the beginning of file(before F character) Type alt-n, then arrow right, then insert a character. Observe that the character is inserted at the beginning of the line (where the alt-n left you).
  1. Repeat steps using <kbd>alt-o</kbd> instead of <kbd>alt-n</kbd>.

  **Expected**: Both alt-o and alt-n behave the same

  **Issues we've seen:**
  bug

- [x] Ubuntu: Keystrokes involving <kbd>ctrl</kbd> resolve to the default layout instead of the active layout (https://github.com/atom/atom/issues/13170)

  <details>
  <summary>Repro steps</summary>

  **This should be tested on Linux with gnome**

  1. Install a Linux ditro with gnome, for example Ubuntu
  1. Install French AZERTY and English US QWERTY layouts
  1. Change the default layout to be French AZERTY
  1. Open Atom and switch to English US QWERTY layout
  1. Press ctrl-w

  **Expected**: `core:close` to dispatch (Or the keybinding-resolver to resolve to ctrl-w). Should be resolving to the keyboard layout that is chosen and not the OS default layout.

  **Issues we've seen:** `core:undo` dispatched because it was resolved as ctrl-z because AZERTY has Z where W is on QWERTY.
  </details>

- [ ] Other keyboard layouts on new Electron version
  - [x] macOS: ABC - Extended
  - [x] macOS: Slovak
  - [x] macOS: U.S. International
  - [ ] macOS: Squirrel (https://github.com/rime/squirrel)
  - [x] macOS: Full-Width Alphanumeric (Google) - Japanese layout  (https://www.google.co.jp/ime/)
  - [x] Windows: U.S.
  - [x] Linux: Pinyin Simplified

### UI

- [x] tree-view drag image (https://github.com/atom/tree-view/pull/1054)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a project in Atom with multiple files and a few folders
  1. Drag a file into a folder

  **Expected:**  
  image

  **Issues we've seen:**

  image
  </details>

- [x] drag-and-drop indicator (https://github.com/atom/tree-view/pull/1055, https://github.com/atom/tabs/pull/426, https://github.com/atom/tabs/pull/437)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a few folder projects on tree-view (you can do this on terminal in a directory that has existing folder and having a space between each folder. Example: `atom folderA folderB folderC`)
  1. Drag and drop all of the folders to reorder them on Atom

  **Expected:**
  The placeholder indicates where the folder/project is being dropped to.

  screen shot 2017-03-29 at 16 31 16

  **Issues we've seen:**
  No placeholder shows up after drag and dropping

  screen shot 2017-03-29 at 16 30 51
  </details>

- [x] Large file rendering (https://github.com/atom/atom/issues/16591)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has over one million lines
  1. Scroll down to the bottom

  **Expected:** To not regress the number of lines that can be rendered. Rendering to be correct for lines past a certain point. Atom 1.25  can render around 800k to 1 million lines correctly.

  **Issues we've seen:** Increased number of lines rendered but bad rendering past a certain point

  bad renderer
  </details>

- [x] Loss of subpixel AA when the cursor is at the end of long lines (https://github.com/atom/atom/issues/16889, https://github.com/atom/atom/pull/16595)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open [text-editor.js](https://github.com/atom/atom/blob/2e19cf4a7c0a4d624b1f9c1072cd292de112cdb0/src/text-editor.js) in Atom
  1. Copy the first 21 lines to a new untitled window
  1. Turn on soft wrapping and change language mode to javascript
  1. Resize the Window so it's soft wrapped
  1. On line 15 type `a` until the you reach the end of the window

  **Expected:** To not lose subpixel AA

  **Issues we've seen:** Loss of subpixel AA. Both when soft wrapping was [enabled](https://github.com/atom/atom/pull/16943) and [disabled](https://github.com/atom/atom/pull/16595).

  subpixel aa loss softwrap
  </details>

- [x] Scrolling horizontally shift + scroll wheel (https://github.com/atom/atom/pull/12696#issuecomment-288615912)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has long lines
  1. Turn off soft wrapping and resize the window so you have a horizontal scrollbar
  1. Press shift and scroll in both directions using the mouse wheel

  **Expected:** File to scroll horizontally in both directions

  **Issues we've seen:** File does not scroll horizontally at all in any direction
  </details>

- [x] Scrollbars misbehaving on the first file that is opened (https://github.com/atom/atom/pull/12696#issuecomment-251299649)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom
  1. Open some file that has content enough to have a vertical scrollbar
  1. Close Atom
  1. Open Atom and look at the vertical scrollbar

  **Expected:** Scrollbar to be visible

  **Issues we've seen:** Scrollbar is not visible and is flickering when you are editing
  </details>

- [x] Middle clicking on unsaved tab (https://github.com/atom/atom/issues/15197)
	- [x] macOS
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom and open a new unsaved file.
  1. Make edits to the unsaved file and middle click on the tab.

  **Expected:**
  Clicking save/cancel or the options on the dialog works.

  **Issues we've seen:**
  The UI and the dialog is unable to receive mouse clicks. You can still choose options via Keyboard, but not mouse.
  </details>

- [x] Linux: Atom scrolls even when not focused (https://github.com/atom/atom/issues/15482)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a large enough file that has a scrollbar
  1. Open a different application and place it over the Atom window
  1. Scroll with the mouse inside the other window
  1. Focus the Atom window

  **Expected:** Atom window should keep the original scroll position

  **Issues we've seen:** The Atom window scrolls after it is focused
  </details>

### Other

- [ ] Deprecation warnings (https://github.com/atom/atom/pull/12696#issuecomment-288871751)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  **This should be tested on macOS, Linux and Windows with community packages installed**

  1. Open Atom
  1. Open the Developer tools console on the master branch and check for deprecation warnings
  1. Open the Developer tools console on the electron upgrade branch and check for deprecation warnings
  1. Compare the two outputs

  **Expected:** No new deprecation warnings

  **Issues we've seen:** New deprecation warnings both from core and community packages
  </details>

- [x] Supported Versions of OS (https://github.com/atom/atom/issues/15297)
	- [x] macOS
	- [x] Windows 10
	- [x] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Check the list of supported platforms [Electron Master](https://github.com/electron/electron/blob/master/docs/tutorial/support.md#supported-platforms)
  1. Check what version is listed as minimum for each platform on https://atom.io and https://atom.io/faq

  **Expected:** https://atom.io and https://atom.io/faq to display the same versions as the Electron documentation

  **Issues we've seen:** macOS version support changed and https://atom.io and https://atom.io/faq was not updated
  </details>

- [x] macOS : Native tabs (https://github.com/atom/atom/pull/14711)
  <details>
  <summary>Repro steps</summary>

  1. Open two atom windows on mac
  1. Go to View in menus and check `Show Tab Bar`
  screen shot 2017-06-03 at 5 10 32 pm
  1. Do this for both Atom windows.
  1. Drag one atom window to the other via tab bar

  **Expected:**
  Native Tab bar shows up and able to drag different windows to each other.  Also make sure UI isn't messed up.

  **Issues we've seen:**
  Native Tab Bar doesn't work and messes up UI when enabled
  </details>

- [x] Ubuntu with KDE: Menu uses 100% CPU (https://github.com/electron/electron/issues/8455, https://github.com/atom/atom/issues/13885)
  <details>
  <summary>Repro steps</summary>

  1. Install Atom on Ubuntu KDE. [[instructions](https://linuxconfig.org/how-to-install-kde-plasma-desktop-on-ubuntu-18-04-bionic-beaver-linux)]
  1. Enable KDE plasma global menu: Right-click on the desktop -> Add Panel => Application Menu Bar. [[demo](https://youtu.be/qo5fhiXAUbI)]
  1. Start Atom.
  1. Try to use the desktop menu.

  **Expected:**
  Desktop menu items work

  **Issues we've seen:**
  Desktop menu items do nothing when you click on things.
  </details>

- [x] macOS: Slovak QWERTZ (https://github.com/atom/atom-keymap/issues/223)
  <details>
  <summary>Repro steps</summary>

  1. Install Slovak keyboard layout on mac
  1. Open keyboard viewer to see the keys are are suppose to show up and open keyboard resolver on atom.
  1. hit `cmd+'`.

  **Expected:**  Resolves to `cmd+'` like it shows on keyboard viewer on mac.

  **Issues we've seen:** It resolves to `ctrl+§`
  </details>

- [x] macOS: IME jump (https://github.com/atom/atom/issues/15696)
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Chinese pinyin
  1. Type one character

  **Expected:**
  Expected the IME to be positioned at under the character you entered
  **Issues we've seen:** The IME window is in the top left corner. When you enter the second character it jumps to be positioned under the text
  </details>

### CODE REVIEW

**Comment by Remagpie** [comment]
The most recent stable version of the electron is [7.1.2](https://www.npmjs.com/package/electron) now.
Is there any reason for picking the old version?

**Comment by darangi** [comment]
@Remagpie, electron upgrades on atom are done one version at a time until the latest version is reached.

**Review by trongham — APPROVED** [review]

**Comment by auvipy** [comment]
one question, why not electron 6/7?

**Comment by darangi** [comment]
> one question, why not electron 6/7?

https://github.com/atom/atom/pull/20172#issuecomment-560128449

**Comment by auvipy** [comment]
OK cool

File: package.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "4.2.7",
+  "electronVersion": "5.0.12",
```
Comment (mfonville) [inline package.json]: 5.0.13 is available too

**Comment by Remagpie** [comment]
Is this PR still being worked on? I'd like to help if there's something I can do.

**Comment by darangi** [comment]
> Is this PR still being worked on? I'd like to help if there's something I can do.

Thanks for offering! 😄 However, this PR is just few failing tests (about 2) from completion.

**Comment by Remagpie** [comment]
You can simplify some of your commits such as https://github.com/atom/atom/pull/20172/commits/e882b8002689a2ba175aa80b0c3ed7963318d47a and https://github.com/atom/atom/pull/20172/commits/6c822302477c1c5919fa3916f21de6f98e5a7e2b.
I think it'd be nice to polish patches a bit so that other people can track history easier.

**Comment by darangi** [comment]
> You can simplify some of your commits such as [e882b80](https://github.com/atom/atom/commit/e882b8002689a2ba175aa80b0c3ed7963318d47a) and [6c82230](https://github.com/atom/atom/commit/6c822302477c1c5919fa3916f21de6f98e5a7e2b).
> I think it'd be nice to polish patches a bit so that other people can track history easier.

Thanks for the recommendation 😃

**Comment by mfonville** [comment]
Congrats on your progress for the Electron-upgrade process. Is there any ETA for which release/milestone of Atom the electron-upgrade will be applied?

**Comment by aminya** [comment]
@darangi Thank you for upgrading. Is there any plan for continuing upgrading Electron to upper versions?
https://www.electronjs.org/blog

**Comment by darangi** [comment]
> @darangi Thank you for upgrading. Is there any plan for continuing upgrading Electron to upper versions?
> https://www.electronjs.org/blog

@aminya Thanks! Yes, we have plans in the future to do that 👍

**Comment by aminya** [comment]
> 
> 
> > @darangi Thank you for upgrading. Is there any plan for continuing upgrading Electron to upper versions?
> > [electronjs.org/blog](https://www.electronjs.org/blog)
> 
> @aminya Thanks! Yes, we have plans in the future to do that 👍

Awesome! That's a very good way to enhance overall Atom's performance.

Would you please take a look at this? Maybe this is related to this PR
https://github.com/atom/atom/issues/20531

**Comment by mstmustisnt** [comment]
Hi,
I'm sorry to bother you, but how can I enable using webview in my Atom plugin? My plugin is using it. 
`By default the webview tag is disabled in Electron >= 5`
https://www.electronjs.org/docs/api/webview-tag#enabling

**Pull request merged**